### PR TITLE
Enable tone generator

### PR DESCRIPTION
**Pull request by ranj063**
This set of patches contains the changes required to enable the tone pipeline. 
Because hostless pipelines, do not have a FE DAI, we introduce the virtual FE that will be created when the siggen widget is loaded. This virtual FE will be used to enable the BE DAI in the case of hostless pipelines when the kcontrol attached to the siggen widget is triggered. 

The patchset also proposes a new switch kcontrol IO handler that will be used for triggering the hostless pipeline.

### CODE REVIEW

File: sound/soc/sof/topology.c [hunk 1]
```diff
@@ -1181,6 +1182,9 @@ static int sof_widget_load_siggen(struct snd_soc_component *scomp, int index,
 	struct sof_ipc_comp_tone tone;
 	struct soc_mixer_control *sm;
 	struct snd_sof_control *scontrol;
+	const char *link_name = "TONE-VFE0";
+	const char *cpu_dai_name = "sof-audio";
```
Comment (ranj063) [inline sound/soc/sof/topology.c]: @plbossart @lgirdwood @keyonjie
With the recent DAI changes, the cpu dai name "sof-audio" doesnt exist anymore.
So I 'll have to change this accordingly. Keyon is helping me with it. I will update the PR based on that

Comment (lgirdwood) [reply sound/soc/sof/topology.c]: Link name should come from topology, i.e it could be constructed from tone widget name/sname.

File: sound/soc/sof/sof-priv.h [hunk 2]
```diff
@@ -221,6 +221,8 @@ struct snd_sof_control {
 	enum sof_ipc_ctrl_cmd cmd;
 	u32 *volume_table; /* volume table computed from tlv data*/
 
+	const char *vfe_link_name; /* name of Virtual FE link */
```
Comment (lgirdwood) [inline sound/soc/sof/sof-priv.h:246]: how do we deal with > 1 VFEs ? Best to have a list.

File: sound/soc/soc-pcm.c [hunk 3]
```diff
@@ -3103,6 +3104,57 @@ static int soc_rtdcom_mmap(struct snd_pcm_substream *substream,
 	return -EINVAL;
 }
 
+/*
+ * for virtual FE dai links, there is no need
+ * to register PCM device. So only allocate memory for
+ * the dummy pcm device and substreams to be used for
+ * codec startup
+ */
+static int setup_virtual_fe(struct snd_soc_pcm_runtime *rtd,
```
Comment (lgirdwood) [inline sound/soc/soc-pcm.c]: iirc, naming convention is like soc_vfe_setup()

File: sound/soc/soc-pcm.c [hunk 4]
```diff
@@ -3103,6 +3104,57 @@ static int soc_rtdcom_mmap(struct snd_pcm_substream *substream,
...
+		return -ENOMEM;
+
+	/* define substream */
+	substream->pcm = pcm;
+	substream->pstr = pstr;
+	substream->number = 0;
+	substream->stream = stream_dir;
+
+	/* prepend substream name with VFE link name to make it unique */
+	sprintf(substream->name, "%s subdevice #%i", rtd->dai_link->name, 0);
```
Comment (lgirdwood) [inline sound/soc/soc-pcm.c]: The dai name and number can hopefully come from topology.

File: sound/soc/soc-pcm.c [hunk 5]
```diff
@@ -2759,6 +2759,8 @@ int soc_dpcm_runtime_update(struct snd_soc_card *card)
 	mutex_unlock(&card->mutex);
 	return 0;
 }
+EXPORT_SYMBOL_GPL(soc_dpcm_runtime_update);
```
Comment (lgirdwood) [inline sound/soc/soc-pcm.c]: The naming may need updated too to add snd_soc prefix. Check that it's also in the public header as well.

File: sound/soc/sof/sof-priv.h (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/sof-priv.h:246]: @lgirdwood but my understanding is that there would be multiple VFE's associated with different kcontrols. But one kcontrol would only ever be associated with only one VFE. Please correct me if I am wrong.

File: sound/soc/soc-pcm.c (hunk 5 above)
Comment (ranj063) [reply sound/soc/soc-pcm.c]: @lgirdwood, this function is in soc-dpcm.h but there's a comment above that says "internal use only". Does that mean it was intended not to be used by modules?

**Comment by ranj063** [comment]
@lgirdwood @plbossart. I've updated the commits with the requested changes.

Please note that some of the new changes depend on the SOFT PR
https://github.com/thesofproject/soft/pull/23

File: sound/soc/soc-pcm.c [hunk 6]
```diff
@@ -1582,7 +1582,6 @@ static int dpcm_add_paths(struct snd_soc_pcm_runtime *fe, int stream,
 
 	/* Create any new FE <--> BE connections */
 	for (i = 0; i < list->num_widgets; i++) {
-
```
Comment (plbossart) [inline sound/soc/soc-pcm.c]: remove change here, not needed

File: sound/soc/soc-pcm.c [hunk 7]
```diff
@@ -2789,6 +2788,115 @@ int soc_dpcm_be_digital_mute(struct snd_soc_pcm_runtime *fe, int mute)
...
+
+	link = kzalloc(sizeof(*link), GFP_KERNEL);
+	if (!link)
+		return -ENOMEM;
+
+	dev_dbg(card->dev, "ASoC: adding new virtual FE DAI link %s\n",
+		link_name);
+
+	/* define FE DAI link */
+	link->name = kstrdup(link_name, GFP_KERNEL);
```
Comment (plbossart) [inline sound/soc/soc-pcm.c:2813]: where are all these link fields freed?

File: sound/soc/soc-pcm.c [hunk 8]
```diff
@@ -2789,6 +2788,115 @@ int soc_dpcm_be_digital_mute(struct snd_soc_pcm_runtime *fe, int mute)
...
+
+	/* enable capture */
+	if (stream_dir == SNDRV_PCM_STREAM_CAPTURE)
+		link->dpcm_capture = 1;
+
+	link->dobj.index = index;
+	link->dobj.type = SND_SOC_DOBJ_DAI_LINK;
+
+	/* add link to card dai link list */
+	snd_soc_add_dai_link(card, link);
```
Comment (plbossart) [inline sound/soc/soc-pcm.c:2839]: can we use snd_soc_remove_dai_link for the free?
I am still worried about this function which isn't symmetrical and has potential to break

File: sound/soc/soc-pcm.c [hunk 9]
```diff
@@ -2789,6 +2788,115 @@ int soc_dpcm_be_digital_mute(struct snd_soc_pcm_runtime *fe, int mute)
...
+				kfree(rtdcom1);
+
+			/* remove dai_link from card */
+			snd_soc_remove_dai_link(card, rtd->dai_link);
+
+			/* free link */
+			kfree(rtd->dai_link);
+
+			/* free runtime */
+			kfree(rtd);
```
Comment (plbossart) [inline sound/soc/soc-pcm.c:2895]: really not hot on this code.

File: sound/soc/sof/control.c [hunk 10]
```diff
@@ -195,3 +210,89 @@ int snd_sof_bytes_put(struct snd_kcontrol *kcontrol,
...
+		(struct soc_mixer_control *)kcontrol->private_value;
+	struct snd_sof_control *scontrol = sm->dobj.private;
+	struct snd_sof_dev *sdev = scontrol->sdev;
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	unsigned int i, channels = scontrol->num_channels;
+
+	pm_runtime_get_sync(sdev->dev);
+
+	/* get all the mixer data from DSP */
+	snd_sof_ipc_get_comp_data(sdev->ipc, scontrol, SOF_IPC_COMP_GET_VALUE,
```
Comment (plbossart) [inline sound/soc/sof/control.c:349]: I still don't get this, why do we need to get the data from DSP? It's not volatile, is it?

File: sound/soc/sof/control.c [hunk 11]
```diff
@@ -195,3 +210,89 @@ int snd_sof_bytes_put(struct snd_kcontrol *kcontrol,
...
+	struct snd_sof_control *scontrol = sm->dobj.private;
+	struct snd_sof_dev *sdev = scontrol->sdev;
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	unsigned int i, channels = scontrol->num_channels;
+	int ret = 0, new_state, old_state;
+	int changed = 0;
+
+	pm_runtime_get_sync(sdev->dev);
+
+	switch (get_widget_type(sdev, scontrol)) {
```
Comment (plbossart) [inline sound/soc/sof/control.c:379]: can you test the result of get_widget_type first and only do the switch if the value is valid (-EINVAL makes no sense here).

File: sound/soc/sof/topology.c [hunk 12]
```diff
@@ -1219,24 +1267,48 @@ static int sof_widget_load_siggen(struct snd_soc_component *scomp, int index,
...
-				  tone.comp.hdr.cmd, &tone, sizeof(tone), r,
-				  sizeof(*r));
+	/*
+	 * create virtual FE link
+	 * This will be used to enable the codec when
+	 * the siggen pipeline is triggered
+	 */
+	ret = snd_soc_dpcm_vfe_new(scomp->card, index, vfe.link_name,
+				   vfe.cpu_dai_name, vfe.platform_name,
+				   SNDRV_PCM_STREAM_PLAYBACK);
```
Comment (plbossart) [inline sound/soc/sof/topology.c:1301]: I think we should make these vfe either playback or capture - with an addition to dapm to mirror the siggen functionality for capture

File: sound/soc/sof/control.c (hunk 10 above)
Comment (ranj063) [reply sound/soc/sof/control.c:349]: @plbossart I am not sure I follow the question. The switch_get() function is just like the volume_get() function to read the values from the DSP. Why do you think it is not necessary to read the switch status values from the DSP? Could you please elaborate?

File: sound/soc/sof/topology.c (hunk 12 above)
Comment (ranj063) [reply sound/soc/sof/topology.c:1301]: The VFE is already defined such that it is playback or capture. The last argument in the vfe_new() function defines the stream direction for the VFE. In the case of siggen, I specify the direction for playback .

I will add a new dapm widget for siggen_sink that will be used for capture.

File: sound/soc/sof/control.c (hunk 10 above)
Comment (plbossart) [reply sound/soc/sof/control.c:349]: the premise for kcontrols is that we only do an IPC if strictly necessary, cause IPC can be slow. This is why we have regmap and caches.
So we should only get the values from the DSP if they are volatile and can change on their own in the firmware. If the driver already has the information, it can provide it back directly.
If you look at the put() case it's exactly what you do, you only send an IPC when the value changed, so why do we need an IPC for get()?

Comment (ranj063) [reply sound/soc/sof/control.c:349]: @plbossart Thanks for the explanation. The put() case was a bit more straightforward to make sure that we send the IPC only when the values are changed. 

But I get what you mean now. The switch status is never changed in the firmware on its own. So it makes sense to just return the value in the driver. I'll update it.

File: sound/soc/soc-pcm.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/soc-pcm.c:2813]: oh yes, let me free them before I free the link. Thanks!

File: sound/soc/soc-pcm.c (hunk 8 above)
Comment (ranj063) [reply sound/soc/soc-pcm.c:2839]: @plbossart I am already using snd_soc_remove_dai_link() in vfe_free(). The problem is that it doesnt free the pcm and its child structure for substream. 

But thinking a bit more about this, I think it makes sense to allocate the pcm and the substream when the vfe link is activated ie in the switch_put() function. That way, I can make the vfe_new() and vfe_free() functions symmertic. 

And the siwtch_put() and switch_get() methods will be symmetric in terms of creating the pcm and substream and freeing them. Let me make the change and update the PR

**Comment by plbossart** [comment]
@ranj063 also if you can add the ALSA/ASoC changes first it helps, you are mixing the framework and SOF changes and it's hard to follow.

**Comment by ghost** [comment]
@ranj063 @plbossart 
Currently I found an issue of the module reload on BYT, which is the kernel panic.
It is related to the rt5651's tplg. the rt5651 use the "get/put/info ID" with 259.
but the kernel has not merged this patch.
which will cause the "Tone Switch 5" kcontol allocate failure.
when we do the module reload, this will cause the kernel panic in the remove_widget() function.

this is the panic log of the kernel:
ops: 0000 [#1] SMP PTI
[   72.434905] Modules linked in: nls_iso8859_1 snd_soc_sst_bytcr_rt5651 intel_soc_dts_thermal intel_soc_dts_iosf intel_powerclamp coretemp asix usbnet kvm_intel kvm irqbypass punit_atom_debug crct10dif_pclmul crc32_pclmul ghash_clmulni_intel sof_acpi_dev(-) pcbc aesni_intel snd_sof_intel_bdw snd_sof_intel_byt aes_x86_64 snd_sof_intel_hsw crypto_simd snd_sof snd_soc_rt5651 cryptd snd_sof_xtensa_dsp glue_helper snd_soc_acpi_intel_match snd_soc_acpi snd_soc_rl6231 joydev input_leds snd_soc_core(E) snd_seq_midi intel_cstate snd_seq_midi_event snd_rawmidi snd_compress ac97_bus snd_seq snd_pcm_dmaengine snd_pcm snd_seq_device mei_txe snd_timer mei snd lpc_ich shpchp soundcore dw_dmac mac_hid dw_dmac_core 8250_dw pwm_lpss_platform spi_pxa2xx_platform pwm_lpss parport_pc ppdev lp parport autofs4 hid_generic
[   72.434997]  usbhid hid mmc_block i915 i2c_algo_bit r8169 drm_kms_helper mii syscopyarea sysfillrect sysimgblt fb_sys_fops drm video sdhci_acpi sdhci
[   72.435027] CPU: 0 PID: 1984 Comm: rmmod Tainted: G            E     4.17.0+ #9
[   72.435031] Hardware name: Circuitco Minnowboard Max B3 PLATFORM/MinnowBoard MAX, BIOS MNW2MAX1.X64.0094.R01.1612052239 12/05/2016
[   72.435061] RIP: 0010:remove_widget+0xaf/0x150 [snd_soc_core]
[   72.435066] RSP: 0018:ffffb9e7c263bbb8 EFLAGS: 00010246
[   72.435072] RAX: ffff908232d919c8 RBX: 0000000000000000 RCX: 0000000000000001
[   72.435077] RDX: 0000000000000001 RSI: fffff29a01c17140 RDI: 0000000040000000
[   72.435081] RBP: ffff90823778c2a8 R08: ffff9082305c5a20 R09: 00000001802a0023
[   72.435086] R10: ffffb9e7c263bae0 R11: ffffffffb9913700 R12: 0000000000000000
[   72.435090] R13: ffff908232e9b000 R14: 0000000000000000 R15: ffff90823778c2a8
[   72.435096] FS:  00007f4cba9a5700(0000) GS:ffff908236c00000(0000) knlGS:0000000000000000
[   72.435100] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   72.435105] CR2: 0000000000000070 CR3: 000000006d386000 CR4: 00000000001006f0
[   72.435109] Call Trace:
[   72.435140]  snd_soc_tplg_component_remove+0x89/0x3b0 [snd_soc_core]
[   72.435155]  snd_sof_free_topology+0x32/0x50 [snd_sof]
[   72.435179]  soc_pcm_private_free+0x4a/0x60 [snd_soc_core]
[   72.435197]  snd_pcm_free+0x1f/0x50 [snd_pcm]
[   72.435214]  __snd_device_free+0x48/0x80 [snd]
[   72.435227]  snd_device_free_all+0x2b/0x50 [snd]
[   72.435240]  release_card_device+0x32/0x90 [snd]
[   72.435253]  device_release+0x30/0x90
[   72.435262]  kobject_put+0x84/0x1b0
[   72.435276]  snd_card_free+0x65/0x90 [snd]
[   72.435299]  snd_soc_unregister_card+0x69/0xe0 [snd_soc_core]
[   72.435308]  release_nodes+0x105/0x1f0
[   72.435317]  device_release_driver_internal+0x172/0x220
[   72.435325]  bus_remove_device+0xe5/0x150
[   72.435332]  device_del+0x138/0x360
[   72.435342]  ? acpi_lpss_platform_notify+0x32/0x180
[   72.435349]  platform_device_del.part.12+0x1e/0x70
[   72.435356]  platform_device_unregister+0x13/0x20
[   72.435364]  sof_acpi_remove+0x1b/0x40 [sof_acpi_dev]
[   72.435371]  platform_drv_remove+0x23/0x40
[   72.435379]  device_release_driver_internal+0x15a/0x220
[   72.435386]  driver_detach+0x39/0x70
[   72.435393]  bus_remove_driver+0x51/0xd0
[   72.435402]  __x64_sys_delete_module+0x13f/0x280
[   72.435411]  do_syscall_64+0x55/0x110
[   72.435420]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[   72.435427] RIP: 0033:0x7f4cba4d7d37
[   72.435431] RSP: 002b:00007ffda87f11d8 EFLAGS: 00000206 ORIG_RAX: 00000000000000b0
[   72.435438] RAX: ffffffffffffffda RBX: 0000000000000000 RCX: 00007f4cba4d7d37
[   72.435442] RDX: 000000000000000a RSI: 0000000000000800 RDI: 0000557e245b9268
[   72.435447] RBP: 0000557e245b9200 R08: 0000000000000000 R09: 1999999999999999
[   72.435451] R10: 0000000000000883 R11: 0000000000000206 R12: 00007ffda87f13f0
[   72.435456] R13: 00007ffda87f1d75 R14: 0000000000000000 R15: 0000557e245b9200
[   72.435461] Code: 10 e8 36 fa 71 f7 48 8b 45 f8 48 85 c0 0f 84 9d 00 00 00 39 5d e8 0f 8e 94 00 00 00 8b 55 20 4c 63 e3 83 fa 01 4e 8b 34 e0 75 af <49> 8b 7e 70 e8 08 fa 71 f7 eb a4 85 c9 7e 76 45 31 f6 4d 63 fe 
[   72.435563] RIP: remove_widget+0xaf/0x150 [snd_soc_core] RSP: ffffb9e7c263bbb8
[   72.435567] CR2: 0000000000000070
[   72.435576] ---[ end trace cd7a507d567e852a ]---

**Comment by ghost** [comment]
@ranj063 @plbossart 
what about adding the kcontrol pointer check in the remove_widget() function?
we could avoid such similarly issue in future, when something is incorrect in topology.

**Comment by lgirdwood** [comment]
@zhigang-wu can you log this as a kerneul bug and send a PR with your fix.

**Comment by plbossart** [comment]
@zhigang-wu it's a good stress test for the kernel, we should detect that an ID is not valid and deal with this case to avoid the crash. I don't want to remove anything for now, just fix the error handling.

**Comment by ghost** [comment]
@lgirdwood @plbossart 
OK, I will log this kernel bug and work for this.

**Comment by plbossart** [comment]
closing, no update or follow-up
